gluten-core/src/main/scala/io/glutenproject/extension/CommonSubexpressionEliminateRule.scala

```diff
@@ -0,0 +1,196 @@
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.sql.extension
+
+import org.apache.spark.internal.Logging
+import org.apache.spark.sql.SparkSession
+import org.apache.spark.sql.catalyst.expressions._
+import org.apache.spark.sql.catalyst.expressions.aggregate.AggregateFunction
+import org.apache.spark.sql.catalyst.plans.logical._
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.internal.SQLConf
+
+import scala.collection.mutable
+
+class CommonSubexpressionEliminateRule(session: SparkSession, conf: SQLConf)
+  extends Rule[LogicalPlan]
+  with Logging {
+
+  private var lastPlan: LogicalPlan = null
+
+  override def apply(plan: LogicalPlan): LogicalPlan = {
+    val newPlan = if (plan.resolved && !plan.fastEquals(lastPlan)) {
+      lastPlan = plan
+      visitPlan(plan)
+    } else {
+      plan
+    }
+    newPlan
+  }
+
+  private case class AliasAndAttribute(alias: Alias, attribute: Attribute)
+
+  private case class RewriteContext(exprs: Seq[Expression], child: LogicalPlan)
+
+  private def visitPlan(plan: LogicalPlan): LogicalPlan = {
+    var newPlan = plan match {
+      case project: Project => visitProject(project)
+      // TODO: CSE in Filter doesn't work for unknown reason, need to fix it later
+      // case filter: Filter => visitFilter(filter)
+      case window: Window => visitWindow(window)
+      case aggregate: Aggregate => visitAggregate(aggregate)
+      case sort: Sort => visitSort(sort)
+      case other =>
+        val children = other.children.map(visitPlan)
+        other.withNewChildren(children)
+    }
+
+    if (newPlan.output.size == plan.output.size) {
+      return newPlan
+    }
+
+    // Add a Project to trim unnecessary attributes(which are always at the end of the output)
+    val postProjectList = newPlan.output.take(plan.output.size)
+    Project(postProjectList, newPlan)
+  }
+
+  private def replaceCommonExprWithAttribute(
+      expr: Expression,
+      commonExprMap: mutable.HashMap[ExpressionEquals, AliasAndAttribute]): Expression = {
+    val exprEquals = commonExprMap.get(ExpressionEquals(expr))
+    if (exprEquals.isDefined) {
+      exprEquals.get.attribute
+    } else {
+      expr.mapChildren(replaceCommonExprWithAttribute(_, commonExprMap))
+    }
+  }
+
+  private def isValidCommonExpr(expr: Expression): Boolean = {
+    if (expr.isInstanceOf[Unevaluable] || expr.isInstanceOf[AggregateFunction]) {
+      return false
+    }
+
+    expr.children.forall(isValidCommonExpr(_))
+  }
+
+  private def rewrite(inputCtx: RewriteContext): RewriteContext = {
+    // scalastyle:off println
```

Comment:
remove all `println`

Reply:
println are replaced with logdebug. They are needed in debug mode